Apps install script for QA

rarescosma opened the pull request:
<!-- Choose your PR title carefully as it will be used as the entry in the changelog! -->

<!-- markdownlint-disable MD041 -->
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [ ] personal data beyond what is necessary for interacting with this pull request, nor
> - [ ] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       <!-- This PR adds a new feature -->
- [ ] kind/improvement   <!-- This PR changes an existing feature -->
- [ ] kind/deprecation   <!-- This PR removes an existing feature -->
- [ ] kind/documentation <!-- This PR contains documentation -->
- [ ] kind/clean-up      <!-- This PR cleans up technical debt -->
- [ ] kind/bug           <!-- This PR fixes a bug -->
- [ ] kind/other         <!-- This PR does something else -->

_Optional_: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change   <!-- This PR introduces an admin facing change, add "Platform Administrator notice" section -->
- [ ] kind/dev-change     <!-- This PR introduces a dev facing change, add "Application Developer notice" section -->
- [ ] kind/security       <!-- This PR introduces a critical security fix, add "Security notice" section -->
- [ ] [kind/adr](set-me\) <!-- This PR implements an ADR, add the link -->

<!-- Uncomment the additional sections that applies. -->

<!-- Additional information to be added in the release notes
### Release notes
...
-->

<!-- Additional information with kind/admin-change
### Platform Administrator notice
...
-->

<!-- Add additional information with kind/dev-change
### Application Developer notice
...
-->

<!-- Add additional information with kind/security
### Security notice
...
-->

### What does this PR do / why do we need this PR?

Implements an installer script for Apps that's as independent as possible from the "install" layer. 

NOTE: This script has been tested *only* on clusters installed with CAPI on the Elastx cloud provider. Should we want to extend support to other installers or cloud provider, we'll do that in separate tickets / PRs.

```
# ./scripts/qa/install_apps.py --help
usage: install_apps.py [-h] -c CONFIG [--sc-subnet SC_SUBNET] [--wc-subnet WC_SUBNET] [-n | --dry-run | --no-dry-run] [-s STEP] [-m | --interactive | --no-interactive]

Configure and install Apps for QA.

options:
  -h, --help            show this help message and exit
  -c, --config CONFIG   secrets configuration file.
  --sc-subnet SC_SUBNET
                        subnet for SC nodes.
  --wc-subnet WC_SUBNET
                        subnet for WC nodes.
  -n, --dry-run, --no-dry-run
                        doesn't perform any effects; just lists steps.
  -s, --step STEP       start at step.
  -m, --interactive, --no-interactive
                        manually confirm each step.
```

Similar to the `ckctl` command series in the Devbox, this script allows skipping to a certain step in the process, manually confirming each step, or dry-running to list the steps. A normal flow looks like this:

```
# ./scripts/qa/install_apps.py -c $CK8S_CONFIG_PATH/qa.json --dry-run
Step 1: Initialize apps
Step 2: Set objectstorage secrets
Step 3: Configure apps
Step 4: Configure secrets
Step 5: Update IPs for SC
Step 6: Install Ingress in SC
Step 7: Configure External DNS for SC
Step 8: Install Dex
Step 9: Update IPs for WC
Step 10: Install Ingress in WC
Step 11: Configure External DNS for WC
Step 12: Sync apps in SC
Step 13: Sync apps in WC
```

The script relies on a minimal `JSON` configuration file that contains the minimum amount of secret material to bootstrap apps in a cluster:

```json
{
  "kubeloginClientSecret": "set-me",
  "awsAccessKey": "set-me",
  "awsSecretKey": "set-me",
  "gcpClientID": "set-me",
  "gcpClientSecret": "set-me",
  "objectStorage": {
    "s3": {
      "accessKey": "set-me",
      "secretKey": "set-me"
    },
    "swift": {
      "applicationCredentialID": "set-me",
      "applicationCredentialSecret": "set-me"
    }
  }
}
```

Credential scope is as follows:

- the AWS credentials are used by the external DNS controller to automatically set up subdomains for both the SC and the WC. These are inferred from the IP address of LoadBalancers associated with the ingress-nginx services, after they stabilized. We do not set any records for the Kubernetes API servers themselves, but this doesn't seem to affect E2E runs.
- the GCP credentials are used to set up the Dex connector in the service cluster. Refer to the [IDP preparation page](https://elastisys.io/welkin/user-guide/prepare-idp/#google) in the Welkin documentation for details. 
- object storage credentials according to your cloud provider
- the `kubeloginClientSecret` should match the value under `.clusters.wc.oidc.client_secret`. This allows us to avoid an additional Control Plane rollout during the Apps install, thus avoiding any interaction with the installer layer.

NOTE: the script assumes that the `$CK8S_CONFIG_PATH/dex-google-group-claim/secret/google-sa-secret.yml` exists and is populated with the proper key (corresponding to the IDP credentials used in the secrets config file)

Fixes https://github.com/elastisys/ck8s-issue-tracker/issues/73

#### Information to reviewers

Prerequisites:
- an empty CAPI cluster on Elastx (with no Apps on it, or on which `./scripts/clean-sc.sh` and `./scripts/clean-wc.sh` have run)
- `$CK8S_CONFIG_PATH/dex-google-group-claim/secret/google-sa-secret.yml` present and client secret configured with the desired callback URL for the domain
- the JSON configuration file populated with the required secrets. A template is available under `scripts/qa/secrets.json.sample`
- `$CK8S_CONFIG_PATH/.state/kube_config_wc.yaml` is initially set to the `kubernetes:admin` configuration. Once the Dex connection is established (after the entire script has run, it can be reverted back to the OIDC config):
  - to get the admin config from CAPI: `./bin/ck8s-capi ops clusterctl mc get kubeconfig <cluster_name> -n capi-cluster > $CK8S_CONFIG_PATH/.state/kube_config_wc.yaml`
  - after the script runs, to get an OIDC-based config: `./bin/ck8s-capi kubeconfig wc`
  
```
./scripts/qa/install_apps.py -c <path_to_secrets_json> --sc-subnet <sc_subnet> --wc-subnet <wc_subnet>
```

#### Checklist

<!-- This section is not added to the changelog or release notes, it is to help you as a contributor and reviewers. -->

- [ ] Proper commit message prefix on all commits
    <!-- Example of commit message prefixes:
    - all: changes to multiple areas
    - apps: changes to applications running in all clusters
    - apps sc: changes to applications running in service clusters
    - apps wc: changes to applications running in workload clusters
    - bin: changes to management binaries
    - config: changes to configuration
    - docs: changes to documentation
    - release: release related
    - scripts: changes to scripts
    - tests: changes to tests
    --->
- Change checks:
    - [ ] The change is transparent
    - [ ] The change is disruptive
    - [ ] The change requires no migration steps
    - [ ] The change requires migration steps
    - [ ] The change updates CRDs
    - [ ] The change updates the config _and_ the schema
- Documentation checks:
    - [ ] The [public documentation](https://github.com/elastisys/welkin) required no updates
    - [ ] The [public documentation](https://github.com/elastisys/welkin) required an update - [link to change](set-me\)
- Metrics checks:
    - [ ] The metrics are still exposed and present in Grafana after the change
    - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts required no updates)
    - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts required an update)
- Logs checks:
    - [ ] The logs do not show any errors after the change
- PodSecurityPolicy checks:
    - [ ] Any changed Pod is covered by Kubernetes Pod Security Standards
    - [ ] Any changed Pod is covered by Gatekeeper Pod Security Policies
    - [ ] The change does not cause any Pods to be blocked by Pod Security Standards or Policies
- NetworkPolicy checks:
    - [ ] Any changed Pod is covered by Network Policies
    - [ ] The change does not cause any dropped packets in the NetworkPolicy Dashboard
- Audit checks:
    - [ ] The change does not cause any unnecessary Kubernetes audit events
    - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
    - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
    - [ ] The bug fix is covered by regression tests

rarescosma commented on scripts/qa/secrets.json.sample:
```diff
@@ -0,0 +1,17 @@
+{
+  "kubeloginClientSecret": "set-me",
+  "awsAccessKey": "set-me",
+  "awsSecretKey": "set-me",
+  "gcpClientID": "set-me",
+  "gcpClientSecret": "set-me",
+  "objectStorage": {
+    "s3": {
+      "accessKey": "set-me",
+      "secretKey": "set-me"
+    },
+    "swift": {
+      "applicationCredentialID": "set-me",
+      "applicationCredentialSecret": "set-me"
+    }
+  }
+}
```
> Also what is the purpose of having kubeloginClientSecret?

We're being clever and pre-configuring CAPI with a pre-generated `kubeloginClientSecret`, then installing Apps using an admin Kubeconfig ([CAPI support PR](https://github.com/elastisys/ck8s-cluster-api/pull/420))

This allows us to avoid the control plane rotation entirely. 
But we will have to swap in the OIDC-based kubeconfig after this installer has run, and before running the E2E suite.
But still, the speed gains + the fact that we don't have to call into the installer layer (tell CAPI to rotate control plane) in the middle of the script, made this approach worthwhile.

rarescosma replied:
Fixed the configuration nesting in 04114d93

Will wait for meeting to talk about internal templating.

rarescosma commented on scripts/qa/install_apps.py:
```diff
@@ -0,0 +1,483 @@
+def install_ingress(cluster: str) -> str:
+    """Install Ingress"""
+    # fmt: off
+    _run_ck8s(
+        "ops", "helmfile", cluster,
+        "-l", "app=admin-namespaces",
+        "-l", "app=dev-namespaces",
+        "-l", "bootstrap=prometheus",
+        "-l", "bootstrap=ingress-nginx",
+        "sync",
+    )
```
Because it pulls in heavy guns like gatekeeper and whatnot. Whereas with granular labels I was able to get away with "just" waiting for prometheus (even tho we only really need the serviceMonitor definition from there).

All I wanted at this stage was the thinnest installation of ingress possible. The sync in the last steps will overwrite it eventually. 

But I'll give your suggestion a go and see what pods / wait time I end up with.

aarnq replied:
Yes, because ingress-nginx is highly privileged and we constrain that privilege with gatekeeper.

I feel like there is a point missing here that this is for QA, so striving away from how things are supposed to work will also make it unsuitable for the task it is implemented for.

rarescosma replied:
Ok, but are we fine with this being in an intermediate step? 

We do a full sync as the last step anyway which will deploy all the ingress constraints.

simonklb commented on scripts/qa/install_apps.py:
```diff
@@ -0,0 +1,483 @@
+
+BIN_DIR: Path = SCRIPT_DIR / ".." / ".." / "bin"
+
+ADMIN_USERS: list[str] = ["admin@example.com", "dev@example.com"]
+ADMIN_GROUP: str = "ck8sdevops@elastisys.com"
+ALL_IPS: dict[str, list[str]] = {"ips": ["0.0.0.0/0"]}
+
+DEV_DOMAIN: str = "dev-ck8s.com"
+DEV_HOSTED_ZONE: str = "Z1001117397DAU71G3RN2"
```
> But I also fear that by adding more and more cloud providers this will grow into a "mini-devbox".

Isn't that necessary though? Otherwise we will only perform QA on ElastX, no?

aarnq commented on scripts/qa/install_apps.py:
```diff
@@ -0,0 +1,483 @@
+        },
+    )
+
+    if args.sc_subnet is not None:
+        sc_config.set(
+            "networkPolicies.global",
+            {
+                "scApiserver": {"ips": [args.sc_subnet]},
+                "scNodes": {"ips": [args.sc_subnet]},
+            },
+        )
+
+    if args.wc_subnet is not None:
+        wc_config.set(
+            "networkPolicies.global",
+            {
+                "wcApiserver": {"ips": [args.wc_subnet]},
+                "wcNodes": {"ips": [args.wc_subnet]},
+            },
+        )
```
So, the one subnet update-ips can pull is the infra subnet with CAPI+OpenStack, but it is missing the ones set for tunnels.
So with a Cluster API env it would easily break, because it doesn't take tunnel IPs into consideration.

Either allow multiple to be specified, split targeting between control plane and worker nodes, or create an issue and attach it to this one that this should be fixed.

elastisys-staffan reviewed (commented):
This is very impressive work! Nice, elegant code. Also, pretty cool that you did it all with the standard library.

That said, I don't feel comfortable giving detailed feedback here. This is because I don't have a good enough understanding (yet) of the Apps deployment flow and the QA process.

I will however give some high-level remarks:

- **Complexity:** Do we really need things like threads and decorator factories for this use case? Maybe we do (and I probably lack the context to really know), but it could't hurt to take a step back and look at the bigger picture. In the end we're "just" applying manifests to Kubernetes clusters here.
- **Speed:** I agree with @aarnq previous comments on this one. If this is intended for QA, we should aim for a vanilla representation that is as close to a real scenario as possible and that means leave as many things in as possible. If we can fully automate the QA process, that speed gain alone is more than enough for our needs.
- **Maintainability:** My concern here has more to do with the original motivation than the PR itself, and is moving into arch territory. Like you said @rarescosma, if we support all providers and an increasing number of use-cases, we risk ending up with a "mini-devbox" here. But I don't have an answer for how we should deal with it.

aarnq commented:
Opened [a PR with some additions](https://github.com/elastisys/compliantkubernetes-apps/pull/2711) (and a reshuffle of things, because I had a hard time following it).

One observation is that due to DNS caching update-ips broke for ingress, and we might not need to address this here and now, but we should probably add some flags to update-ips to skip ingress, so we can run that after we know DNS is setup.

Second observation which is only really applicable to Kubespray clusters is that we should not run update-ips on WC until we have Dex up.
As Kubespray uses Ansible to run kubectl commands from the nodes we can and do create OIDC kubeconfigs that don't work until Dex is up. Preparing certificate based kubeconfigs would be an additional step.

rarescosma commented:
> Opened [a PR with some additions](https://github.com/elastisys/compliantkubernetes-apps/pull/2711) (and a reshuffle of things, because I had a hard time following it).

Merged it, and also reworked the configuration part into a separate module. 

It's a runnable module:
- running `./scripts/qa/config.py` will give you a "maximal" configuration example (where each possible leaf in the config tree is expanded)
- `./scripts/qa/config.py --minimal` will output the minimally viable config

I've also removed the intermediary "update-ips" call and broke the dependency between the "install ingress" and "install DNS" steps, so the flow can be started with the DNS step now, without getting an empty value for the ingress IP.

elastisys-staffan commented:
I just did an "idiot test" meaning that I put an idiot (i.e me) at the helm and try to run the script with as little effort as possible on Brewer.

There were a few bumps in the road but they were all due to either underlying quirks in ck8s/ck8s-capi or mistakes made by me.

Aside from those bumps, I've got apps up and running with ease and basically everything I've tested so far has been working! Great work! :tada: 

The only things I've been missing yet is: 
- S3 bucket creation. Maybe it's out of scope for this script but I feel it should at least be included in the prerequisites.
- More extensive documentation. Some examples would be nice in the help section.
- `scSubnets|wcSubnets` are all about network policies, right? They don't map to the equivalents keys in `common-config.yaml|sc-config.yaml|wc-config.yaml`. We don't necessarily have to have the same path (`networkPolicies.global...`) but nevertheless I think it could be made less confusing.

elastisys-staffan reviewed (approved):
I think this is robust enough so I'll approve in case you want to merge before the end of the sprint.

rarescosma commented:
Added a [readme file](https://github.com/elastisys/compliantkubernetes-apps/blob/rares/qa-install-script/scripts/qa/README.md) for the installer script, with sample configurations for Elastx, Upcloud and Brewer.
